# 5. Funciones de Ventana para el Análisis de Datos

In [1]:
# @title Postgress preloading (install server)
#The output of the installation is not displayed when %%capture is used at the start of the cell
%%capture
# Install postgresql server
!sudo apt-get -y -qq update
!sudo apt-get -y -qq install postgresql
!sudo service postgresql start
# Setup a password `postgres` for username `postgres`
!sudo -u postgres psql -U postgres -c "ALTER USER postgres PASSWORD 'postgres';"
# Setup a database with name `sampledb` to be used
# !sudo -u postgres psql -U postgres -c 'DROP DATABASE IF EXISTS sampledb;'
# !sudo -u postgres psql -U postgres -c 'CREATE DATABASE sampledb;'

!curl https://raw.githubusercontent.com/limspiga/data-modeling/main/db/data.dump  -O
!sudo -u postgres psql -U postgres -c 'CREATE DATABASE sqlda;'

# !sudo -u postgres psql -U postgres -d 'sqlda' -f 'data.dump'

# https://thivyapriyaa.medium.com/setting-up-postgresql-on-google-colab-4d02166939fc


In [2]:
# @title Postgress preloading
# import
!sudo -u postgres psql -d sqlda < data.dump
%env DATABASE_URL=postgresql://postgres:postgres@localhost:5432/sqlda
#To load the sql extention to start using %%sql
%load_ext sql

SET
SET
SET
SET
SET
 set_config 
------------
 
(1 row)

SET
SET
SET
SET
CREATE EXTENSION
COMMENT
CREATE EXTENSION
COMMENT
CREATE TEXT SEARCH DICTIONARY
ALTER TEXT SEARCH DICTIONARY
SET
SET
CREATE TABLE
ALTER TABLE
CREATE TABLE
ALTER TABLE
CREATE TABLE
ALTER TABLE
CREATE MATERIALIZED VIEW
ALTER TABLE
CREATE TABLE
ALTER TABLE
CREATE MATERIALIZED VIEW
ALTER TABLE
CREATE TABLE
ALTER TABLE
CREATE TABLE
ALTER TABLE
CREATE TABLE
ALTER TABLE
CREATE TABLE
ALTER TABLE
CREATE TABLE
ALTER TABLE
CREATE TABLE
ALTER TABLE
CREATE TABLE
ALTER TABLE
CREATE TABLE
ALTER TABLE
COPY 44533
COPY 0
COPY 50000
COPY 32
COPY 50000
COPY 20
COPY 418158
COPY 12
COPY 15412
COPY 37711
COPY 300
COPY 20
ALTER TABLE
ALTER TABLE
CREATE INDEX
CREATE INDEX
CREATE INDEX
CREATE INDEX
CREATE INDEX
CREATE INDEX
CREATE INDEX
CREATE INDEX
REFRESH MATERIALIZED VIEW
REFRESH MATERIALIZED VIEW
env: DATABASE_URL=postgresql://postgres:postgres@localhost:5432/sqlda



En el Capítulo 3, exploraste diversas funciones de SQL esenciales para la preparación de datos, incluyendo CASE WHEN, COALESCE y NULLIF, las cuales se aplican a datos específicos dentro de una fila individual. En contraste, el Capítulo 4 te guió a través de funciones de agregación, tales como SUM, AVG y COUNT, diseñadas para procesar y analizar múltiples filas de datos. Imagina, por ejemplo, que estás analizando indicadores de salud en recién nacidos; podrías emplear CASE WHEN para evaluar la salud de cada infante basándote en su peso y altura, mientras que las funciones de agregación te permitirían calcular promedios y desviaciones estándar a nivel nacional.

En ciertas situaciones, la posición relativa de un dato dentro de un conjunto adquiere una importancia crítica, como es el caso de los rankings. Por ejemplo, un bebé podría estar en un rango de salud diferente comparado a nivel nacional versus a nivel municipal. Además, al manejar conjuntos de datos extensos, es posible que necesites analizar subgrupos o particiones específicas, como podrían ser los diferentes estados. En estos casos, se hace necesario calcular un ranking que sea válido únicamente dentro de cada partición, tomando en cuenta un conjunto predefinido de filas, o "ventana". Este método de análisis se realiza mediante el uso de las llamadas funciones de ventana.


# Funciones de Ventana
Continuando con el análisis de las Funciones de Ventana, supongamos que tu objetivo es identificar los clientes pioneros de ZoomZoom. En términos técnicos, tu intención es
- ordenar a los clientes basándote en la fecha en que iniciaron su relación con la empresa,
- asignando al más antiguo el rango 1, al siguiente el rango 2, y así consecutivamente.

Para lograrlo, puedes recopilar la información de todos los clientes mediante la consulta que se presenta a continuación:

In [ ]:
%%sql
-- Figure 5.1: Customers ordered by date_added
SELECT
  customer_id, first_name, last_name, date_added
FROM
  customers
ORDER BY
  date_added LIMIT 10;

10 rows affected.


[(18685, 'Ingram', 'Crossman', datetime.datetime(2012, 11, 9, 0, 0)),
 (30555, 'Stanwood', 'Stookes', datetime.datetime(2012, 11, 9, 0, 0)),
 (17099, 'Pearla', 'Halksworth', datetime.datetime(2012, 11, 9, 0, 0)),
 (6173, 'Danila', 'Gristwood', datetime.datetime(2012, 11, 9, 0, 0)),
 (12484, 'Lillis', 'Brayley', datetime.datetime(2012, 11, 9, 0, 0)),
 (30046, 'Nanete', 'Hassur', datetime.datetime(2012, 11, 9, 0, 0)),
 (13390, 'Danika', 'Lough', datetime.datetime(2012, 11, 9, 0, 0)),
 (7486, 'Ciro', 'Ferencowicz', datetime.datetime(2012, 11, 9, 0, 0)),
 (2625, 'Binky', 'Dawtrey', datetime.datetime(2012, 11, 9, 0, 0)),
 (35683, 'Betteanne', 'Rulf', datetime.datetime(2012, 11, 9, 0, 0))]

Puedes
- ordenar a los clientes desde el más antiguo hasta el más reciente,
- copiar los resultados en una hoja de Excel y
- asignar un número de fila a cada uno para tener el rango de cada cliente.

Sin embargo, este método no es automático y puede llevar a errores. SQL ofrece varias formas para lograr esto de manera más eficiente.

En secciones posteriores de este capítulo, descubrirás cómo asignar números secuenciales a registros previamente ordenados utilizando la función de ventana RANK. Como punto de partida, puedes emplear una función de agregación para recopilar las fechas de registro de los clientes y proceder a su ordenación.


In [ ]:
%%sql
-- Figure 5.2: Aggregate date-time ordering
SELECT
  date_added, COUNT(*)
FROM
  customers
GROUP BY
  date_added
ORDER BY
  date_added LIMIT 10;

 * postgresql://postgres:***@localhost:5432/sqlda
10 rows affected.


[(datetime.datetime(2012, 11, 9, 0, 0), 11),
 (datetime.datetime(2012, 11, 10, 0, 0), 13),
 (datetime.datetime(2012, 11, 11, 0, 0), 12),
 (datetime.datetime(2012, 11, 12, 0, 0), 19),
 (datetime.datetime(2012, 11, 13, 0, 0), 23),
 (datetime.datetime(2012, 11, 14, 0, 0), 16),
 (datetime.datetime(2012, 11, 15, 0, 0), 20),
 (datetime.datetime(2012, 11, 16, 0, 0), 14),
 (datetime.datetime(2012, 11, 17, 0, 0), 11),
 (datetime.datetime(2012, 11, 18, 0, 0), 21)]

El resultado ordena las fechas de manera secuencial. A partir de ello, es posible determinar cuántos clientes se unieron a ZoomZoom antes que un cliente específico, sumando las inscripciones de los días anteriores a su fecha de adhesión. Sin embargo, este enfoque es manual y no ofrece una visión directa del rango. Es aquí donde entran en juego las funciones de ventana. Estas herramientas procesan varias filas de datos conservando toda su información. Son esenciales para tareas como determinar rangos. En la siguiente sección, ilustraremos con un ejemplo de consulta utilizando una función de ventana.

# Conceptos Básicos de las Funciones de Ventana
A continuación, se presenta la sintaxis básica de una función de ventana:

```sql
SELECT {columns},
{window_func} OVER (PARTITION BY {partition_key} ORDER BY {order_key})
FROM table1;
```

Aquí, {columnas} son las columnas que deseas recuperar de las tablas para la consulta, {func_ventana} es la función de ventana que deseas usar, table1 es la tabla o tablas unidas de las que deseas extraer datos, y la palabra clave OVER indica donde comienza la definición de la ventana. Esta definición, en su sintaxis básica, incluye dos partes: {clave_partición} y {clave_orden}. La primera se refiere a la columna o columnas que deseas dividir (se hablará más de esto más adelante) y la segunda es la columna o columnas por las que deseas ordenar.

Para ilustrar esto, consideremos un ejemplo. Podrías pensar que no conoces ninguna función de ventana, pero la realidad es que todas las funciones de agregación pueden utilizarse como funciones de ventana. Ahora, utiliza COUNT(*) en la siguiente consulta:

In [ ]:
%%sql
-- Figure 5.3: Customers listed using the COUNT(*) window query
SELECT
  customer_id,
  title,
  first_name,
  last_name,
  gender,
  COUNT(*) OVER () as total_customers
FROM
  customers
ORDER BY
  customer_id
LIMIT 10;

 * postgresql://postgres:***@localhost:5432/sqlda
10 rows affected.


[(1, None, 'Arlena', 'Riveles', 'F', 50000),
 (2, 'Dr', 'Ode', 'Stovin', 'M', 50000),
 (3, None, 'Braden', 'Jordan', 'M', 50000),
 (4, None, 'Jessika', 'Nussen', 'F', 50000),
 (5, None, 'Lonnie', 'Rembaud', 'F', 50000),
 (6, None, 'Cortie', 'Locksley', 'M', 50000),
 (7, None, 'Wood', 'Kennham', 'M', 50000),
 (8, None, 'Rutger', 'Humblestone', 'M', 50000),
 (9, None, 'Melantha', 'Tibb', 'F', 50000),
 (10, 'Ms', 'Barbara-anne', 'Gowlett', 'F', 50000)]

Como se muestra en la Figura 5.3, la consulta devuelve título, primer nombre y apellido, al igual que una consulta SELECT típica. Sin embargo, ahora hay una nueva columna llamada total_clientes. Esta columna contiene el recuento de usuarios que serían creados por la siguiente consulta:

In [ ]:
%%sql
SELECT
  COUNT(*)
FROM
  customers;

 * postgresql://postgres:***@localhost:5432/sqlda
1 rows affected.


[(50000,)]

La consulta anterior devuelve 50,000. La consulta retornó todas las filas y el COUNT(*) en la consulta devuelve el recuento como lo haría cualquier función de agregación normal.
Ahora, en cuanto a los otros parámetros de la consulta, ¿qué sucede si añades la cláusula OVER para convertir este COUNT en una función de ventana, manteniendo la función como COUNT pero definiendo la ventana usando PARTITION BY, como en la consulta siguiente?

In [ ]:
%%sql
-- Figure 5.4: Customers listed using COUNT(*) partitioned by the gender window query
SELECT
  customer_id,
  title,
  first_name,
  last_name,
  gender,
  COUNT(*) OVER (PARTITION BY gender) as total_customers
FROM
  customers
ORDER BY
  customer_id
LIMIT 10;

 * postgresql://postgres:***@localhost:5432/sqlda
10 rows affected.


[(1, None, 'Arlena', 'Riveles', 'F', 25044),
 (2, 'Dr', 'Ode', 'Stovin', 'M', 24956),
 (3, None, 'Braden', 'Jordan', 'M', 24956),
 (4, None, 'Jessika', 'Nussen', 'F', 25044),
 (5, None, 'Lonnie', 'Rembaud', 'F', 25044),
 (6, None, 'Cortie', 'Locksley', 'M', 24956),
 (7, None, 'Wood', 'Kennham', 'M', 24956),
 (8, None, 'Rutger', 'Humblestone', 'M', 24956),
 (9, None, 'Melantha', 'Tibb', 'F', 25044),
 (10, 'Ms', 'Barbara-anne', 'Gowlett', 'F', 25044)]

Aquí, puedes ver que total_clientes ahora ha cambiado los conteos a uno de dos valores, 24956 o 25044. Al usar la cláusula PARTITION BY sobre la columna de género, SQL divide el conjunto de datos en múltiples particiones basadas en los valores únicos de esta columna. Dentro de cada partición, SQL calcula el conteo total. Por ejemplo, hay 24956 hombres, por lo que la función de ventana COUNT para la partición de hombres devuelve 24596, lo que puedes confirmar con la siguiente consulta:

In [ ]:
%%sql
SELECT
  gender,
  COUNT(*)
FROM
  customers
GROUP BY
1;

 * postgresql://postgres:***@localhost:5432/sqlda
2 rows affected.


[('M', 24956), ('F', 25044)]

Ahora ves cómo se define y se utiliza la partición con la cláusula PARTITION BY. Para las mujeres, el conteo es igual al número de mujeres, y para los hombres, el conteo es igual al número de hombres. ¿Qué sucede ahora si utilizas ORDER BY en lugar de ello en la cláusula OVER como sigue?

In [ ]:
%%sql
-- Figure 5.5: Customers listed using COUNT(*) ordered by the customer_id window query
SELECT
  customer_id, title,
  first_name, last_name, gender,
  COUNT(*) OVER (ORDER BY customer_id) as total_customers
FROM
  customers
ORDER BY
  customer_id
LIMIT 10;

 * postgresql://postgres:***@localhost:5432/sqlda
10 rows affected.


[(1, None, 'Arlena', 'Riveles', 'F', 1),
 (2, 'Dr', 'Ode', 'Stovin', 'M', 2),
 (3, None, 'Braden', 'Jordan', 'M', 3),
 (4, None, 'Jessika', 'Nussen', 'F', 4),
 (5, None, 'Lonnie', 'Rembaud', 'F', 5),
 (6, None, 'Cortie', 'Locksley', 'M', 6),
 (7, None, 'Wood', 'Kennham', 'M', 7),
 (8, None, 'Rutger', 'Humblestone', 'M', 8),
 (9, None, 'Melantha', 'Tibb', 'F', 9),
 (10, 'Ms', 'Barbara-anne', 'Gowlett', 'F', 10)]

Notarás algo similar a un conteo acumulado para el total de clientes. Aquí es donde proviene la definición de 'ventana' en función de ventana. Cuando usas esta función de ventana, dado que no especificaste un PARTITION BY, se utiliza el conjunto de datos completo para el cálculo. Dentro de este conjunto de datos,
- cuando no se especifica ORDER BY, se asume que solo hay una ventana, que contiene todo el conjunto de datos.
- Sin embargo, cuando se especifica ORDER BY, las filas en el conjunto de datos se ordenan de acuerdo con este criterio.

Para cada valor único en el orden, SQL forma un grupo de valores, que contiene todas las filas con dicho valor. La consulta luego crea una ventana para cada grupo de valores.

- **La ventana contendrá todas las filas en este grupo de valores y todas las filas que están ordenadas antes de este grupo de valores**.

A continuación se muestra un ejemplo:

<figure>
<center><img src='https://github.com/limspiga/data-modeling/blob/main/images/8daed639-c93e-4a0c-a7d4-27c16ee9c892.png?raw=true' width="600" />
<figcaption>
Figure 5.6: Windows for customers using COUNT(*) ordered by the customer_id window query</figcaption></center>
</figure>

Aquí, el conjunto de datos se ordena usando customer_id, que resulta ser la clave primaria. Como tal, cada fila tiene un valor único y forma un grupo de valores. El primer grupo de valores, sin ninguna fila antes que él, forma su propia ventana, que contiene solo la primera fila. La ventana del segundo grupo de valores contendrá tanto a sí misma como a la fila anterior, lo que significa la primera y segunda fila. Luego, la ventana del tercer grupo de valores contendrá a sí misma y las dos filas anteriores, y así sucesivamente. Cada grupo de valores tiene su ventana. Una vez que se establecen las ventanas, para cada grupo de valores, se calcula la función de ventana basada en la ventana. En este ejemplo, esto significa que COUNT se aplica a cada ventana. Por lo tanto, el grupo de valores 1 (la primera fila) obtiene 1 como resultado ya que su Ventana 1 contiene una fila, el grupo de valores 2 (la segunda fila) obtiene 2 ya que su Ventana 2 contiene dos filas, y así sucesivamente. Los resultados se aplican a cada fila en este grupo de valores si el grupo contiene varias filas. Ten en cuenta que la ventana se utiliza solo para el cálculo. Los resultados se asignan a filas en el grupo de valores, no se asignan a las filas en la ventana.

¿Qué sucede cuando combinas PARTITION BY y ORDER BY? Ahora, observa la siguiente consulta:

In [ ]:
%%sql
-- Figura 5.7: Clientes listados usando COUNT(*) particionados por género y ordenados por la consulta de ventana customer_id.
SELECT
  customer_id,
  title,
  first_name,
  last_name,
  gender,
  COUNT(*) OVER (
    PARTITION BY gender
    ORDER BY customer_id
  ) as total_customers
FROM
  customers
ORDER BY
  customer_id
LIMIT 10;

 * postgresql://postgres:***@localhost:5432/sqlda
10 rows affected.


[(1, None, 'Arlena', 'Riveles', 'F', 1),
 (2, 'Dr', 'Ode', 'Stovin', 'M', 1),
 (3, None, 'Braden', 'Jordan', 'M', 2),
 (4, None, 'Jessika', 'Nussen', 'F', 2),
 (5, None, 'Lonnie', 'Rembaud', 'F', 3),
 (6, None, 'Cortie', 'Locksley', 'M', 3),
 (7, None, 'Wood', 'Kennham', 'M', 4),
 (8, None, 'Rutger', 'Humblestone', 'M', 5),
 (9, None, 'Melantha', 'Tibb', 'F', 4),
 (10, 'Ms', 'Barbara-anne', 'Gowlett', 'F', 5)]

Al igual que la consulta anterior, parece ser una especie de rango. Sin embargo, parece diferir según el género. En este SQL en particular, la consulta divide la tabla en dos subconjuntos basados en la columna PARTITION BY. Eso es porque la cláusula PARTITION BY, al igual que GROUP BY, primero divide el conjunto de datos en grupos (que aquí se llama partición) basados en el valor en la columna de género. Luego, cada partición se utiliza como base para hacer un conteo, con cada partición teniendo su propio conjunto de grupos de valores. Estos grupos de valores se ordenan dentro de la partición, se crean ventanas basadas en los grupos de valores y sus órdenes, y la función de ventana se aplica a los valores. Los resultados se asignan finalmente a cada fila en los grupos de valores.


Este proceso se ilustra en la Figura 5.8. Este proceso produce el conteo que puedes ver. Las tres palabras clave, OVER(), PARTITION BY y ORDER BY, son la base del poder de las funciones de ventana.

<figure>
<center><img src='https://github.com/limspiga/data-modeling/blob/main/images/6a82dd5c-88c2-4fec-bd7e-228233ed5585.png?raw=true' width="600" />
<figcaption>
Figura 5.8: Ventanas para clientes listados usando COUNT(*) particionados por género y ordenados por la consulta de ventana customer_id.</figcaption></center>
</figure>

Ahora que comprendes las funciones de ventana, intenta aplicarlas en el próximo ejercicio.

## Ejercicio 5.01: Analizando tasas de llenado de datos de clientes a lo largo del tiempo.

En este ejercicio, aplicarás funciones de ventana a un conjunto de datos y analizarás la información.
- Durante los últimos seis meses, ZoomZoom ha estado experimentando con diversas promociones para lograr que sus clientes estén más involucrados en la actividad de venta.

- Una forma de medir el nivel de compromiso es evaluar la disposición de las personas para completar todos los campos en el formulario de cliente, especialmente su dirección. Para lograr este objetivo, la empresa desea un total acumulado de cuántos usuarios han rellenado sus direcciones postales con el tiempo.

Realiza los siguientes pasos para completar este ejercicio:

1. Utiliza funciones de ventana y escribe una consulta que devuelva la información de los clientes y cuántas personas han completado su dirección de calle. Además, ordena la lista por fecha.

In [ ]:
%sql SELECT   customer_id, street_address, date_added::DATE FROM customers WHERE customer_id IN (2625, 30555, 48307, 13390, 12484) LIMIT 4;

 * postgresql://postgres:***@localhost:5432/sqlda
4 rows affected.


[(2625, '0353 Iowa Road', datetime.date(2012, 11, 9)),
 (12484, None, datetime.date(2012, 11, 9)),
 (13390, '38463 Forest Dale Way', datetime.date(2012, 11, 9)),
 (30555, '294 Quincy Hill', datetime.date(2012, 11, 9))]

In [ ]:
%%sql
-- La ventana es por día (10/11)
-- Figura 5.9: Filtro de dirección de calle ordenado por la consulta de ventana date_added
SELECT
  customer_id,
  street_address,
  date_added::DATE,

  -- not_null_street_address
  COUNT(
    CASE
      WHEN street_address IS NOT NULL THEN customer_id
      ELSE NULL -- LO HACEMOS NULO PARA QUE NO LO CUENTE
    END
  ) OVER (ORDER BY date_added::DATE) AS not_null_street_address,

  -- total_street_address
  COUNT(*) OVER (ORDER BY date_added::DATE) as total_street_address
FROM
  customers
 ORDER BY
 date_added LIMIT 10;

 * postgresql://postgres:***@localhost:5432/sqlda
10 rows affected.


[(6173, '79865 Hagan Terrace', datetime.date(2012, 11, 9), 10, 11),
 (7486, '61 Village Crossing', datetime.date(2012, 11, 9), 10, 11),
 (13390, '38463 Forest Dale Way', datetime.date(2012, 11, 9), 10, 11),
 (2625, '0353 Iowa Road', datetime.date(2012, 11, 9), 10, 11),
 (30046, '13961 Steensland Trail', datetime.date(2012, 11, 9), 10, 11),
 (17099, '130 Marcy Crossing', datetime.date(2012, 11, 9), 10, 11),
 (12484, None, datetime.date(2012, 11, 9), 10, 11),
 (18685, '86 Michigan Junction', datetime.date(2012, 11, 9), 10, 11),
 (48307, '8487 Warbler Plaza', datetime.date(2012, 11, 9), 10, 11),
 (30555, '294 Quincy Hill', datetime.date(2012, 11, 9), 10, 11)]


3. Escribe una consulta para ver cómo cambian con el tiempo el número de personas que completan el campo de la dirección de calle.
4. En el paso 1, ya obtuviste todas las direcciones de los clientes ordenadas por la fecha de registro. En el query 5.10, las dos columnas que siguen a la columna de fecha de registro son el número de direcciones no NULL y el número de todas las direcciones de los clientes para cada día acumulativo, es decir, una suma desde el inicio de las ventas hasta el día actual.
Como aprendiste en el Capítulo 4, "Funciones de Agregación para Análisis de Datos", al dividir el número de direcciones no NULL entre el número de todas las direcciones de los clientes, puedes obtener el porcentaje de clientes con direcciones de calle no NULL y derivar el porcentaje de clientes con direcciones de calle NULL. Seguir este número proporcionará una visión de cómo los clientes interactúan con tu fuerza de ventas a lo largo del tiempo. Además, dado que ambos números de direcciones se calculan para cada día acumulativo, el porcentaje también es para cada día acumulativo. Este es un ejemplo de diferentes funciones de ventana compartiendo la misma ventana en la misma consulta.

  También puedes reescribir la siguiente consulta utilizando una cláusula WINDOW para simplificarla, lo cual se introducirá en la próxima sección.

In [ ]:
%%sql
WITH
  -- FUNCION WINDOWS
  daily_rolling_count as (
    SELECT
      customer_id,
      street_address,
      date_added::DATE,
      COUNT(
        CASE
          WHEN street_address IS NOT NULL THEN customer_id
          ELSE NULL
        END
      ) OVER (ORDER BY date_added::DATE)
        as non_null_street_address,
      COUNT(*) OVER (ORDER BY date_added::DATE)
        as total_street_address

    FROM
      customers
  )

SELECT DISTINCT
  date_added,
  non_null_street_address,
  total_street_address,
  1 - 1.0 * non_null_street_address/total_street_address
    AS null_address_percentage
FROM
  daily_rolling_count
ORDER BY
  date_added
LIMIT 10;

 * postgresql://postgres:***@localhost:5432/sqlda
10 rows affected.


[(datetime.date(2012, 11, 9), 10, 11, Decimal('0.09090909090909090909')),
 (datetime.date(2012, 11, 10), 22, 24, Decimal('0.08333333333333333333')),
 (datetime.date(2012, 11, 11), 33, 36, Decimal('0.08333333333333333333')),
 (datetime.date(2012, 11, 12), 50, 55, Decimal('0.09090909090909090909')),
 (datetime.date(2012, 11, 13), 71, 78, Decimal('0.08974358974358974359')),
 (datetime.date(2012, 11, 14), 86, 94, Decimal('0.08510638297872340426')),
 (datetime.date(2012, 11, 15), 105, 114, Decimal('0.07894736842105263158')),
 (datetime.date(2012, 11, 16), 118, 128, Decimal('0.07812500000000000000')),
 (datetime.date(2012, 11, 17), 128, 139, Decimal('0.07913669064748201439')),
 (datetime.date(2012, 11, 18), 145, 160, Decimal('0.09375000000000000000'))]

Este resultado te proporcionará la lista del porcentaje acumulativo de direcciones de calle NULL para cada día. Luego, puedes proporcionar el conjunto completo de datos a software de análisis de datos y visualización, como Excel, para estudiar la tendencia general de los datos, descubrir patrones de cambio y hacer sugerencias sobre cómo aumentar la participación de los clientes para la dirección de la empresa.

En este ejercicio, has aprendido cómo usar funciones de ventana para analizar datos. En la siguiente sección, comprenderás cómo utilizar la palabra clave WINDOW en tus consultas.

## La Palabra Clave WINDOW
Ahora que comprendes los conceptos básicos de las funciones de ventana, es hora de introducir una sintaxis que facilitará su escritura. En muchos escenarios, tu análisis implica ejecutar múltiples funciones contra la misma ventana para que puedas compararlos uno al lado del otro, y es muy probable que los ejecutes dentro de la misma consulta. Por ejemplo, cuando estás realizando un análisis basado en género, es posible que desees calcular un total acumulado de clientes y también un total acumulado de clientes con un título, utilizando la misma partición que se basa en el género. Esto te llevará a escribir la siguiente consulta:

In [ ]:
%sql SELECT title FROM customers WHERE customer_id = 2;

 * postgresql://postgres:***@localhost:5432/sqlda
1 rows affected.


[('Dr',)]

In [ ]:
%%sql
-- Figura 5.11: Total acumulado de clientes en general con el título por consulta de ventana de género
SELECT
  customer_id,
  title,
  first_name,
  last_name,
  gender,
  COUNT(*) OVER (
    PARTITION BY gender ORDER BY customer_id  -- OVER 1
  ) as total_customers, -- CUENTA EL TOTAL DE CLIENTES POR GENERO

  SUM(CASE WHEN title IS NOT NULL THEN 1 ELSE 0 END) OVER (. -- COMVIERTE EN BINARIO EL TITLE PARA CONTAR
   PARTITION BY gender ORDER BY customer_id -- OVER 2
  ) as total_customers_title

FROM customers
ORDER BY customer_id
LIMIT 10;

 * postgresql://postgres:***@localhost:5432/sqlda
10 rows affected.


[(1, None, 'Arlena', 'Riveles', 'F', 1, 0),
 (2, 'Dr', 'Ode', 'Stovin', 'M', 1, 1),
 (3, None, 'Braden', 'Jordan', 'M', 2, 1),
 (4, None, 'Jessika', 'Nussen', 'F', 2, 0),
 (5, None, 'Lonnie', 'Rembaud', 'F', 3, 0),
 (6, None, 'Cortie', 'Locksley', 'M', 3, 1),
 (7, None, 'Wood', 'Kennham', 'M', 4, 1),
 (8, None, 'Rutger', 'Humblestone', 'M', 5, 1),
 (9, None, 'Melantha', 'Tibb', 'F', 4, 0),
 (10, 'Ms', 'Barbara-anne', 'Gowlett', 'F', 5, 1)]

Aunque la consulta te proporciona el resultado, puede ser tedioso de escribir, especialmente la cláusula OVER, ya que es la misma para las dos funciones. Afortunadamente, puedes simplificar esto utilizando la cláusula WINDOW para definir una ventana genérica para varias funciones en la misma consulta. La cláusula WINDOW facilita la asignación de un alias a una ventana.
Puedes simplificar la consulta anterior escribiéndola de la siguiente manera:

In [ ]:
%%sql
SELECT
  customer_id,
  title,
  first_name,
  last_name,
  gender,
  COUNT(*) OVER w as total_customers,
  SUM(
CASE
             WHEN title IS NOT NULL THEN 1
ELSE 0 END
  ) OVER w as total_customers_title
FROM
  customers

-- EXTRAEMOS EL OVER Y LO METEMOS EN LA VENTANA W
WINDOW w AS (
  PARTITION BY gender ORDER BY customer_id
)

ORDER BY customer_id
LIMIT 10;

 * postgresql://postgres:***@localhost:5432/sqlda
10 rows affected.


[(1, None, 'Arlena', 'Riveles', 'F', 1, 0),
 (2, 'Dr', 'Ode', 'Stovin', 'M', 1, 1),
 (3, None, 'Braden', 'Jordan', 'M', 2, 1),
 (4, None, 'Jessika', 'Nussen', 'F', 2, 0),
 (5, None, 'Lonnie', 'Rembaud', 'F', 3, 0),
 (6, None, 'Cortie', 'Locksley', 'M', 3, 1),
 (7, None, 'Wood', 'Kennham', 'M', 4, 1),
 (8, None, 'Rutger', 'Humblestone', 'M', 5, 1),
 (9, None, 'Melantha', 'Tibb', 'F', 4, 0),
 (10, 'Ms', 'Barbara-anne', 'Gowlett', 'F', 5, 1)]

Esta consulta debería proporcionarte el mismo resultado que puedes ver en la captura de pantalla anterior. Sin embargo, no tuviste que escribir una larga consulta PARTITION BY y ORDER BY para cada función de ventana. En su lugar, simplemente creaste un alias con la ventana definida como "w".

# Estadísticas con Funciones de Ventana
Ahora que comprendes cómo funcionan las funciones de ventana, puedes empezar a usarlas para calcular estadísticas útiles, como rangos, percentiles y estadísticas acumulativas.
En la siguiente tabla, se han resumido una variedad de funciones estadísticas que son útiles. También es importante enfatizar nuevamente que todas las funciones de agregación también pueden utilizarse como funciones de ventana (AVG, SUM, COUNT, y así sucesivamente):

<figure>
<center>

| Name         | Descripción                                                            |
|--------------|------------------------------------------------------------------------|
| ROW_NUMBER   | Numera la fila actual dentro de su partición empezando desde 1.        |
| DENSE_RANK   | Ordena la fila actual dentro de su partición sin espacios.             |
| RANK         | Ordena la fila actual dentro de su partición con espacios.             |
| LAG          | Devuelve un valor evaluado en la fila que está a una distancia específica antes de la fila actual dentro de la partición. |
| LEAD         | Devuelve un valor evaluado en la fila que está a una distancia específica después de la fila actual dentro de la partición. |
| NTILE        | Divide las filas en una partición tan igualmente como sea posible y asigna a cada fila un entero comenzando desde 1 hasta el valor del argumento. |

<figcaption>
Figura 5.12: Funciones estadísticas de ventana.</figcaption></center>
</figure>



Normalmente, una llamada a cualquiera de estas funciones dentro de una declaración SQL irá seguida de la palabra clave OVER. Luego, esta palabra clave será seguida por más palabras clave como PARTITION BY y ORDER BY, siendo cualquiera de ellas opcional dependiendo de la función que estés utilizando.
Por ejemplo, la función ROW_NUMBER() se verá de la siguiente manera:


```sql
ROW_NUMBER () OVER (
PARTITION BY column 1, column 2
ORDER BY column_ 3, column_4
)

```

## Ejercicio 5.02: Clasificación de la Contratación
ZoomZoom desea realizar una campaña de marketing para sus clientes más antiguos en diferentes estados.
- ZoomZoom quiere que escribas una consulta que clasifique a los clientes según su fecha de incorporación (date_added) para cada estado.


1. Clasifica cada cliente, donde 1 se asignará a la primera fecha de incorporación, 2 a la segunda y así sucesivamente, utilizando la función RANK():


In [ ]:
%%sql
SELECT
  customer_id,
  first_name,
  last_name,
  state,
  date_added::DATE,
  ROW_NUMBER() OVER (
  -- NTILE(4) OVER ( -- NTILE en SQL es una función de ventana que distribuye
  -- LAG(customer_id, 1)  OVER (
  -- ROW_NUMBER() OVER (
    PARTITION BY state ORDER BY date_added
  ) AS cust_rank
FROM
  customers
ORDER BY
  state, cust_rank
LIMIT 190;

 * postgresql://postgres:***@localhost:5432/sqlda
190 rows affected.


[(10732, 'Dael', 'Persent', 'AK', datetime.date(2012, 11, 12), 1),
 (35117, 'Jourdain', 'Burnyeat', 'AK', datetime.date(2012, 11, 15), 2),
 (6500, 'Eugenia', 'Iacobacci', 'AK', datetime.date(2012, 11, 22), 3),
 (12603, 'Carce', 'Hubber', 'AK', datetime.date(2012, 11, 23), 4),
 (33836, 'Martie', 'Drinnan', 'AK', datetime.date(2012, 12, 23), 5),
 (18621, 'Pammi', 'Manby', 'AK', datetime.date(2013, 1, 5), 6),
 (45170, 'Dunstan', 'Gooddie', 'AK', datetime.date(2013, 1, 19), 7),
 (35168, 'Tracey', 'Ramstead', 'AK', datetime.date(2013, 1, 20), 8),
 (42141, 'Carlin', 'Hammant', 'AK', datetime.date(2013, 2, 24), 9),
 (14407, 'Cleopatra', 'Ferneyhough', 'AK', datetime.date(2013, 3, 11), 10),
 (15148, 'Audre', 'Shepherd', 'AK', datetime.date(2013, 4, 7), 11),
 (9683, 'Frankie', 'Cowderoy', 'AK', datetime.date(2013, 5, 18), 12),
 (25645, 'Jolee', 'Ferres', 'AK', datetime.date(2013, 6, 7), 13),
 (48023, 'Mable', 'Sowrah', 'AK', datetime.date(2013, 6, 15), 14),
 (31531, 'Kendre', 'Whyberd', 'AK', datetime.date(2013, 7, 7), 15),
 (6946, 'Frederik', 'Bilbrooke', 'AK', datetime.date(2013, 7, 8), 16),
 (31688, 'Jehanna', 'Couper', 'AK', datetime.date(2013, 9, 5), 17),
 (15830, 'Lorne', 'Shurman', 'AK', datetime.date(2013, 9, 18), 18),
 (26169, 'Alberta', 'Leguay', 'AK', datetime.date(2013, 9, 22), 19),
 (18099, 'Bevon', 'Dancy', 'AK', datetime.date(2013, 9, 24), 20),
 (31382, 'Andie', 'Barthram', 'AK', datetime.date(2013, 10, 24), 21),
 (2300, 'Jeanette', 'Gladdis', 'AK', datetime.date(2013, 10, 30), 22),
 (4275, 'Shantee', 'Hammerberger', 'AK', datetime.date(2013, 11, 11), 23),
 (43805, 'Rachael', 'Foxworthy', 'AK', datetime.date(2013, 12, 2), 24),
 (44861, 'Chrissy', 'Coffey', 'AK', datetime.date(2014, 1, 6), 25),
 (27491, 'Tim', 'Bridel', 'AK', datetime.date(2014, 1, 7), 26),
 (861, 'Wayland', 'Rivett', 'AK', datetime.date(2014, 2, 19), 27),
 (18225, 'Armand', 'Simonelli', 'AK', datetime.date(2014, 2, 21), 28),
 (1750, 'Sarena', 'Leist', 'AK', datetime.date(2014, 2, 25), 29),
 (40295, 'Eryn', 'Dowse', 'AK', datetime.date(2014, 3, 22), 30),
 (24302, 'Woodman', 'Emeny', 'AK', datetime.date(2014, 4, 8), 31),
 (22551, 'Kameko', 'McDermid', 'AK', datetime.date(2014, 4, 16), 32),
 (36577, 'Melvyn', 'Teideman', 'AK', datetime.date(2014, 5, 5), 33),
 (43509, 'Colette', 'Ivakhno', 'AK', datetime.date(2014, 5, 24), 34),
 (45049, 'Dickie', 'Peto', 'AK', datetime.date(2014, 6, 8), 35),
 (38434, 'Brandice', 'Berzin', 'AK', datetime.date(2014, 6, 24), 36),
 (43913, 'Charmian', 'Struys', 'AK', datetime.date(2014, 7, 11), 37),
 (35248, 'Isiahi', 'Ortas', 'AK', datetime.date(2014, 7, 13), 38),
 (7301, 'Beck', 'Chezier', 'AK', datetime.date(2014, 8, 9), 39),
 (1053, 'Hyman', 'Crowdy', 'AK', datetime.date(2014, 9, 4), 40),
 (37838, 'Kinnie', 'Woodings', 'AK', datetime.date(2014, 9, 10), 41),
 (29615, 'Eileen', 'Messenger', 'AK', datetime.date(2014, 10, 10), 42),
 (43773, 'Maria', 'Kidwell', 'AK', datetime.date(2014, 10, 10), 43),
 (21729, 'Wilfrid', 'Gotch', 'AK', datetime.date(2014, 11, 19), 44),
 (19032, 'Giorgi', 'Skayman', 'AK', datetime.date(2015, 1, 9), 45),
 (18638, 'Maryjane', 'Grant', 'AK', datetime.date(2015, 1, 20), 46),
 (46614, 'Tasha', 'Simoncelli', 'AK', datetime.date(2015, 1, 31), 47),
 (12673, 'Zelig', 'Andersson', 'AK', datetime.date(2015, 2, 1), 48),
 (30732, 'Johny', 'Wolfendell', 'AK', datetime.date(2015, 2, 12), 49),
 (40054, 'Brose', 'Heller', 'AK', datetime.date(2015, 2, 14), 50),
 (38967, 'Adey', 'Quickenden', 'AK', datetime.date(2015, 2, 20), 51),
 (19944, 'Illa', 'Hayball', 'AK', datetime.date(2015, 2, 27), 52),
 (37900, 'Teresa', 'Potebury', 'AK', datetime.date(2015, 2, 28), 53),
 (23794, 'June', 'Rigglesford', 'AK', datetime.date(2015, 3, 13), 54),
 (26360, 'Gian', 'Arnason', 'AK', datetime.date(2015, 4, 2), 55),
 (18783, 'Faith', 'Kubec', 'AK', datetime.date(2015, 4, 2), 56),
 (4902, 'Felecia', 'Brinkworth', 'AK', datetime.date(2015, 4, 22), 57),
 (6931, 'Evelyn', 'Pycock', 'AK', datetime.date(2015, 5, 3), 58),
 (28200, 'Alisha',

Aquí puedes ver a cada cliente con su información y rango en la columna "cust_rank" basado en su fecha de incorporación para cada estado.

En este ejercicio, utilizaste la función RANK() para clasificar los datos en un conjunto de datos en un orden específico. En la próxima sección, aprenderás cómo utilizar el Window Frame

# Window Frame

Como se mencionó en las secciones anteriores donde se discuten los fundamentos de las funciones de ventana, por defecto, se establece una ventana para cada grupo de valores que abarca todas las filas desde la primera hasta la fila actual en la partición, como se muestra en la Figura 5.6. Sin embargo, esto es lo predeterminado y se puede ajustar utilizando la cláusula de intervalo de ventana. Una consulta de función de ventana que utiliza la cláusula de intervalo de ventana se vería de la siguiente manera:


```sql
SELECT
  {columns},
  {window_func} OVER (
    PARTITION BY {partition_key}
    ORDER BY {order_key}
    {rangeorrows} BETWEEN {frame_start} AND {frame_end}
) FROM
{table1};
```



Aquí, {columnas} son las columnas que deseas recuperar de las tablas para la consulta, {función_de_ventana} es la función de ventana que deseas utilizar, {clave_de_partición} es la columna o columnas en las que deseas particionar, {clave_de_orden} es la columna o columnas por las que deseas ordenar, {rango_o_filas} es ya sea la palabra clave RANGE o la palabra clave ROWS, {inicio_del_marco} es una palabra clave que indica dónde comenzar el marco de ventana, {fin_del_marco} es una palabra clave que indica dónde terminar el marco de ventana y {tabla1} es la tabla o tablas unidas de las que deseas obtener datos.

Es importante considerar los posibles valores de {frame_start} y {frame_end}. Estos pueden ser:
- **UNBOUNDED PRECEDING**: Para {frame_start}, hace referencia al primer registro de la partición.
- **{offset} PRECEDING**: Indica {offset} filas o rangos antes de la fila actual.
- **CURRENT ROW**: Corresponde a la fila actual.
- **{offset} FOLLOWING**: Señala {offset} filas o rangos después de la fila actual.
- **UNBOUNDED FOLLOWING**: Para {frame_end}, se refiere al último registro de la partición.

Al ajustar la ventana, podemos derivar estadísticas valiosas como el **promedio móvil**. Este se refiere al promedio de una métrica en un intervalo de tiempo específico. Si, por ejemplo, quieres obtener el promedio móvil de ventas de siete días para ZoomZoom, comienza sumando las ventas diarias usando SUM y agrupándolas por `sales_transaction_date`. Esto generará una lista de ventas por día. Al ordenar la lista por fecha, tendrás una ventana móvil de siete días tomando la fila actual y las seis anteriores. Calculando el AVG de estas siete entradas, obtendrás el promedio móvil deseado para ese día.

La consulta para realizar esta operación es la siguiente:

In [ ]:
%%sql
WITH
  daily_sales as (
    SELECT
      sales_transaction_date::DATE,
      SUM(sales_amount) as total_sales
    FROM sales
    GROUP BY 1
  ),
  moving_average_calculation_7 AS (
    SELECT
      sales_transaction_date,
      total_sales,
      AVG(total_sales) OVER (
        ORDER BY sales_transaction_date
        ROWS BETWEEN 6 PRECEDING and CURRENT ROW -- USAMOS UN WINDOWS FRAME
      ) AS sales_moving_average_7,
      ROW_NUMBER() OVER (
        ORDER BY sales_transaction_date
      ) as row_number
    FROM
      daily_sales
    ORDER BY 1
  )
SELECT
  sales_transaction_date,
  CASE
    WHEN row_number>=7 THEN sales_moving_average_7
    ELSE NULL
  END AS sales_moving_average_7
FROM
  moving_average_calculation_7
  LIMIT 10;

 * postgresql://postgres:***@localhost:5432/sqlda
10 rows affected.


[(datetime.date(2012, 11, 4), None),
 (datetime.date(2012, 11, 6), None),
 (datetime.date(2012, 11, 9), None),
 (datetime.date(2012, 11, 11), None),
 (datetime.date(2012, 11, 12), None),
 (datetime.date(2012, 11, 13), None),
 (datetime.date(2012, 11, 15), 394.2758571428571),
 (datetime.date(2012, 11, 17), 399.99000000000007),
 (datetime.date(2012, 11, 18), 399.99000000000007),
 (datetime.date(2012, 11, 19), 399.99000000000007)]

Al trabajar con una ventana móvil de N días, surge la duda sobre cómo tratar los primeros N-1 días en la columna ordenada. En la consulta que vimos, las primeras seis entradas se marcan como nulas mediante una declaración CASE. Esto se debe a que el promedio móvil de siete días requiere datos de siete días completos. Sin esa declaración, la ventana calcularía promedios para los primeros días basándose en los datos disponibles. Así, el promedio móvil para el segundo día sería la media entre ese día y el anterior, y para el sexto día, el promedio de esos seis días. Ambos enfoques, ya sea usando nulos o promediando los días disponibles, pueden ser válidos según el contexto. Será tarea del analista decidir cuál es más adecuado según la situación específica.

# Ejercicio 5.03: Motivación para el Almuerzo en Equipo
En esta tarea, emplearás un marco de ventana para extraer información valiosa de tus datos.

Con el objetivo de impulsar las ventas, el equipo ha optado por ofrecer un almuerzo a todos los vendedores cada vez que superen el récord de ganancias diarias de los últimos 30 días. Necesitas una consulta que muestre las ventas totales en dólares de un día específico y la meta de ventas que deben superar ese día, a partir del 1 de enero de 2019.

1. Calcula las ventas totales para un día dado

In [5]:
%%sql
WITH
  daily_sales as (
    SELECT
      sales_transaction_date::DATE,
      SUM(sales_amount) as total_sales
    FROM
      sales
    GROUP BY
      1
  ),
  sales_stats_30 AS (
    SELECT
      sales_transaction_date,
      total_sales,
      MAX(total_sales) OVER (
        ORDER BY sales_transaction_date
        ROWS BETWEEN 30 PRECEDING and 1 PRECEDING
      ) AS max_sales_30
    FROM
      daily_sales
    ORDER BY
      1
  )
SELECT
  sales_transaction_date,
  to_char(total_sales, 'FM999G999G999D99') total_sales,
  to_char(max_sales_30, 'FM999G999G999D99') max_sales_30
FROM
  sales_stats_30
WHERE
  sales_transaction_date>='2019-01-01'
LIMIT 40;

 * postgresql://postgres:***@localhost:5432/sqlda
40 rows affected.


[(datetime.date(2019, 1, 1), '6,249.87', '546,899.89'),
 (datetime.date(2019, 1, 2), '119,599.91', '546,899.89'),
 (datetime.date(2019, 1, 3), '71,799.87', '546,899.89'),
 (datetime.date(2019, 1, 4), '130,649.88', '546,899.89'),
 (datetime.date(2019, 1, 5), '6,649.87', '546,899.89'),
 (datetime.date(2019, 1, 6), '73,049.85', '546,899.89'),
 (datetime.date(2019, 1, 7), '122,049.86', '546,899.89'),
 (datetime.date(2019, 1, 8), '197,749.84', '546,899.89'),
 (datetime.date(2019, 1, 9), '5,399.89', '546,899.89'),
 (datetime.date(2019, 1, 10), '252,899.86', '546,899.89'),
 (datetime.date(2019, 1, 11), '164,349.86', '546,899.89'),
 (datetime.date(2019, 1, 12), '215,599.83', '546,899.89'),
 (datetime.date(2019, 1, 13), '348,899.92', '546,899.89'),
 (datetime.date(2019, 1, 14), '274,749.92', '546,899.89'),
 (datetime.date(2019, 1, 15), '136,149.9', '546,899.89'),
 (datetime.date(2019, 1, 16), '5,699.89', '546,899.89'),
 (datetime.date(2019, 1, 17), '203,299.86', '546,899.89'),
 (datetime.date(2019, 1, 18), '119,849.9', '546,899.89'),
 (datetime.date(2019, 1, 19), '356,499.83', '546,899.89'),
 (datetime.date(2019, 1, 20), '163,249.85', '546,899.89'),
 (datetime.date(2019, 1, 21), '213,499.87', '546,899.89'),
 (datetime.date(2019, 1, 22), '351,649.92', '546,899.89'),
 (datetime.date(2019, 1, 23), '303,199.9', '546,899.89'),
 (datetime.date(2019, 1, 24), '278,199.85', '546,899.89'),
 (datetime.date(2019, 1, 25), '190,349.86', '546,899.89'),
 (datetime.date(2019, 1, 26), '237,149.91', '546,899.89'),
 (datetime.date(2019, 1, 27), '334,949.83', '546,899.89'),
 (datetime.date(2019, 1, 28), '300,299.9', '546,899.89'),
 (datetime.date(2019, 1, 29), '74,349.82', '546,899.89'),
 (datetime.date(2019, 1, 30), '357,099.82', '546,899.89'),
 (datetime.date(2019, 1, 31), '75,049.81', '357,099.82'),
 (datetime.date(2019, 2, 1), '405,149.9', '357,099.82'),
 (datetime.date(2019, 2, 2), '258,149.76', '405,149.9'),
 (datetime.date(2019, 2, 3), '514,299.76', '405,149.9'),
 (datetime.date(2019, 2, 4), '357,549.75', '514,299.76'),
 (datetime.date(2019, 2, 5), '290,049.85', '514,299.76'),
 (datetime.date(2019, 2, 6), '121,199.88', '514,299.76'),
 (datetime.date(2019, 2, 7), '122,599.85', '514,299.76'),
 (datetime.date(2019, 2, 8), '411,099.85', '514,299.76'),
 (datetime.date(2019, 2, 9), '60,599.84', '514,299.76')]

Observa el uso de un marco de ventana que va desde 30 PRECEDING hasta 1 PRECEDING. Al utilizar 1 PRECEDING, estás excluyendo la fila actual del cálculo. El resultado es un máximo móvil de 30 días en los 30 días anteriores al día actual.

2. Ahora calcularás las ventas totales de cada día y las compararás con el objetivo de ese día, que es el promedio móvil de 30 días que calculaste en el paso anterior. Las ventas totales de cada día ya se han calculado en el SQL anterior en la primera expresión de tabla común y se hacen referencia más adelante en la consulta principal. Por lo tanto, puedes escribir el siguiente SQL:

In [ ]:
%%sql
WITH
  daily_sales as (
    SELECT
      sales_transaction_date::DATE,
      SUM(sales_amount) as total_sales
    FROM sales
    GROUP BY 1
  ),
  sales_stats_30 AS (
    SELECT
      sales_transaction_date,
      total_sales,
      MAX(total_sales) OVER (
        ORDER BY sales_transaction_date
        ROWS BETWEEN 30 PRECEDING and 1 PRECEDING
      ) AS max_sales_30
    FROM
      daily_sales
    ORDER BY 1
  )
SELECT
  sales_transaction_date,
  to_char(total_sales, 'FM999G999G999D99') total_sales,
  to_char(max_sales_30, 'FM999G999G999D99') max_sales_30
FROM
  sales_stats_30
WHERE
  total_sales > max_sales_30 -- DIFERENCia
AND
  sales_transaction_date>='2019-01-01';



 * postgresql://postgres:***@localhost:5432/sqlda
29 rows affected.


[(datetime.date(2019, 2, 1), '405,149.9', '357,099.82'),
 (datetime.date(2019, 2, 3), '514,299.76', '405,149.9'),
 (datetime.date(2019, 3, 10), '574,449.84', '484,599.83'),
 (datetime.date(2019, 7, 12), '356,129.79', '304,959.83'),
 (datetime.date(2019, 8, 31), '338,629.91', '245,659.88'),
 (datetime.date(2019, 10, 22), '268,354.9', '237,889.85'),
 (datetime.date(2019, 10, 27), '470,234.88', '268,354.9'),
 (datetime.date(2019, 12, 27), '393,599.77', '344,879.81'),
 (datetime.date(2019, 12, 28), '657,459.82', '393,599.77'),
 (datetime.date(2020, 2, 8), '740,254.8', '387,994.82'),
 (datetime.date(2020, 5, 6), '411,539.8', '374,689.79'),
 (datetime.date(2020, 5, 13), '440,719.75', '411,539.8'),
 (datetime.date(2020, 6, 16), '423,499.84', '352,679.79'),
 (datetime.date(2020, 8, 11), '364,079.84', '308,389.82'),
 (datetime.date(2020, 9, 15), '335,149.86', '309,989.86'),
 (datetime.date(2020, 9, 29), '367,519.85', '335,149.86'),
 (datetime.date(2020, 10, 26), '434,739.86', '367,519.85'),
 (datetime.date(2020, 12, 1), '458,759.78', '337,559.86'),
 (datetime.date(2020, 12, 10), '688,559.78', '458,759.78'),
 (datetime.date(2021, 1, 20), '557,229.68', '513,254.79'),
 (datetime.date(2021, 2, 27), '637,369.74', '536,034.79'),
 (datetime.date(2021, 4, 13), '567,019.71', '490,459.82'),
 (datetime.date(2021, 6, 12), '418,959.82', '392,479.83'),
 (datetime.date(2021, 6, 27), '498,889.81', '418,959.82'),
 (datetime.date(2021, 10, 16), '247,094.88', '233,649.8'),
 (datetime.date(2021, 10, 23), '264,444.85', '247,094.88'),
 (datetime.date(2021, 11, 5), '287,079.76', '264,444.85'),
 (datetime.date(2021, 11, 24), '379,414.8', '287,079.76'),
 (datetime.date(2021, 11, 28), '445,259.71', '379,414.8')]

Como puedes ver, los marcos de ventana facilitan el cálculo de estadísticas móviles y, incluso, pueden resultar divertidos. Ahora, concluirás este capítulo con una actividad que pondrá a prueba tu capacidad para utilizar funciones de ventana.

# Actividad 5.01: Análisis de Ventas Utilizando Marcos de Ventana y Funciones de Ventana
En esta tarea, aplicarás funciones de windows y frames de variadas formas para analizar los datos de ventas de ZoomZoom.

Al inicio del año, es esencial definir la estrategia de ventas para el próximo periodo.
El equipo quiere comprender el desempeño global de la empresa y el rendimiento específico de cada día del año. Por ello, el Jefe de Ventas te ha solicitado llevar a cabo este análisis.

1. Calcula el monto total de ventas por día para todos los días del año 2021 (es decir, antes del 1 de enero de 2022).


In [6]:
%%sql
SELECT
  sales_transaction_date::date,
  SUM(sales_amount) sales_amount
FROM sales
WHERE
  sales_transaction_date::date BETWEEN '20210101' AND '20211231'
GROUP BY
  sales_transaction_date::date LIMIT 10;

 * postgresql://postgres:***@localhost:5432/sqlda
10 rows affected.


[(datetime.date(2021, 1, 1), 320739.70100000006),
 (datetime.date(2021, 1, 2), 248239.73599999998),
 (datetime.date(2021, 1, 3), 213419.72799999994),
 (datetime.date(2021, 1, 4), 378919.7169999999),
 (datetime.date(2021, 1, 5), 331084.6939999999),
 (datetime.date(2021, 1, 6), 90089.649),
 (datetime.date(2021, 1, 7), 426449.7289999998),
 (datetime.date(2021, 1, 8), 183449.789),
 (datetime.date(2021, 1, 9), 513254.7929999999),
 (datetime.date(2021, 1, 10), 307994.72699999984)]



2. Calcula el promedio móvil de 30 días para el monto total de ventas diarias.


In [ ]:
%%sql
WITH
  daily_sales as (
    SELECT
      sales_transaction_date::date,
      SUM(sales_amount) sales_amount
    FROM
      sales
    WHERE
      sales_transaction_date::date BETWEEN '20210101' AND '20211231'
    GROUP BY
      sales_transaction_date::date
  )
SELECT
  sales_transaction_date,
  sales_amount,
  AVG(sales_amount) OVER w AS moving_avg
FROM
  daily_sales
WINDOW w AS (
  ORDER BY sales_transaction_date
  ROWS BETWEEN 30 PRECEDING AND 1 PRECEDING
)
ORDER BY 1 LIMIT 10;

 * postgresql://postgres:***@localhost:5432/sqlda
10 rows affected.


[(datetime.date(2021, 1, 1), 320739.70100000006, None),
 (datetime.date(2021, 1, 2), 248239.73599999998, 320739.70100000006),
 (datetime.date(2021, 1, 3), 213419.72799999994, 284489.7185),
 (datetime.date(2021, 1, 4), 378919.7169999999, 260799.72166666668),
 (datetime.date(2021, 1, 5), 331084.6939999999, 290329.7205),
 (datetime.date(2021, 1, 6), 90089.649, 298480.7152),
 (datetime.date(2021, 1, 7), 426449.7289999998, 263748.8708333333),
 (datetime.date(2021, 1, 8), 183449.789, 286991.85057142854),
 (datetime.date(2021, 1, 9), 513254.7929999999, 274049.092875),
 (datetime.date(2021, 1, 10), 307994.72699999984, 300627.50399999996)]

3. Calcula en qué décil estaría cada fecha en comparación con otros días en función del monto total de ventas diarias en el promedio móvil de 30 días.

<figure>
<center><img src='https://github.com/limspiga/data-modeling/blob/main/images/8a310b4d-6704-43b5-aeae-9b5a5d721847.png?raw=true' width="600" />
<figcaption>
Figura 5.12: Funciones de ventana estadísticas.</figcaption></center>
</figure>


Figura 5.17: Déciles para el monto de ventas del concesionario
En esta actividad, utilizaste funciones de ventana para obtener la tendencia de ventas de todo tu año y utilizaste esta tendencia de ventas para identificar los días en los que ZoomZoom está funcionando bien o no tan bien.

In [ ]:
%%sql
WITH
  daily_sales as (
    SELECT
      sales_transaction_date::date,
      SUM(sales_amount) sales_amount
    FROM
      sales
    WHERE
      sales_transaction_date::date BETWEEN '20210101' AND '20211231'
    GROUP BY
      sales_transaction_date::date
  ),
  moving_avg AS (
    SELECT
      sales_transaction_date,
      sales_amount,
      AVG(sales_amount) OVER w AS moving_avg
    FROM
      daily_sales
    WINDOW w AS (
      ORDER BY sales_transaction_date
      ROWS BETWEEN 30 PRECEDING AND 1 PRECEDING
    )
  )
SELECT
  sales_transaction_date,
  sales_amount,
  moving_avg,
  -- PostgreSQL no tiene una función DECILE, pero cuenta con una función
  -- más general llamada NTILE() que puedes utilizar. NTILE(10)
  -- es el equivalente de DECILE().
  NTILE(10) OVER (ORDER BY moving_avg DESC) AS decile
FROM
  moving_avg m
WHERE
  moving_avg IS NOT NULL
ORDER BY
  decile;

 * postgresql://postgres:***@localhost:5432/sqlda
364 rows affected.


[(datetime.date(2021, 1, 2), 248239.73599999998, 320739.70100000006, 1),
 (datetime.date(2021, 1, 12), 164189.76699999996, 313294.72081818176, 1),
 (datetime.date(2021, 1, 11), 432599.66599999997, 301364.2263, 1),
 (datetime.date(2021, 1, 13), 12829.790999999997, 300869.30799999996, 1),
 (datetime.date(2021, 1, 10), 307994.72699999984, 300627.50399999996, 1),
 (datetime.date(2021, 1, 6), 90089.649, 298480.7152, 1),
 (datetime.date(2021, 2, 6), 50389.734999999986, 293352.5650333333, 1),
 (datetime.date(2021, 1, 5), 331084.6939999999, 290329.7205, 1),
 (datetime.date(2021, 2, 2), 161009.56200000003, 288809.9029333333, 1),
 (datetime.date(2021, 2, 3), 275539.70199999993, 287062.8974, 1),
 (datetime.date(2021, 1, 8), 183449.789, 286991.85057142854, 1),
 (datetime.date(2021, 2, 5), 312554.685, 285937.06383333326, 1),
 (datetime.date(2021, 1, 3), 213419.72799999994, 284489.7185, 1),
 (datetime.date(2021, 2, 4), 400689.7019999998, 283616.89689999993, 1),
 (datetime.date(2021, 1, 29), 229079.76699999996, 281014.7344285714, 1),
 (datetime.date(2021, 2, 7), 12709.779999999999, 280817.23189999996, 1),
 (datetime.date(2021, 3, 19), 261699.6079999999, 279505.3776666666, 1),
 (datetime.date(2021, 1, 30), 253264.66699999996, 279223.8734827586, 1),
 (datetime.date(2021, 2, 1), 536034.7929999998, 279216.73436666664, 1),
 (datetime.date(2021, 1, 14), 108059.76200000003, 278712.4220769231, 1),
 (datetime.date(2021, 1, 31), 346484.73399999994, 278358.56659999996, 1),
 (datetime.date(2021, 3, 16), 297969.60299999977, 277742.2141666666, 1),
 (datetime.date(2021, 3, 18), 248179.79299999998, 277530.87763333326, 1),
 (datetime.date(2021, 3, 17), 217469.71499999997, 276911.20853333327, 1),
 (datetime.date(2021, 1, 16), 268244.8109999999, 276875.0686, 1),
 (datetime.date(2021, 3, 20), 89799.68500000007, 276830.87389999995, 1),
 (datetime.date(2021, 3, 15), 242904.73399999997, 276778.04596666654, 1),
 (datetime.date(2021, 1, 17), 16939.724000000002, 276335.6775, 1),
 (datetime.date(2021, 2, 18), 117149.71300000006, 276195.89063333324, 1),
 (datetime.date(2021, 3, 21), 238979.68999999997, 275919.2062999999, 1),
 (datetime.date(2021, 3, 13), 16749.719999999998, 275677.3758999999, 1),
 (datetime.date(2021, 1, 21), 96669.81400000004, 275438.73449999996, 1),
 (datetime.date(2021, 2, 8), 433819.78699999995, 275125.8982666666, 1),
 (datetime.date(2021, 1, 9), 513254.7929999999, 274049.092875, 1),
 (datetime.date(2021, 1, 28), 471704.7319999998, 273952.1419259259, 1),
 (datetime.date(2021, 2, 17), 341934.72099999984, 273869.39113333327, 1),
 (datetime.date(2021, 2, 14), 322899.77199999994, 273744.89476666664, 1),
 (datetime.date(2021, 1, 26), 139389.72999999998, 273044.13651999994, 2),
 (datetime.date(2021, 2, 26), 208989.61999999988, 272681.7213666666, 2),
 (datetime.date(2021, 2, 9), 81829.72700000001, 272478.06473333325, 2),
 (datetime.date(2021, 2, 19), 355599.74500000005, 272064.39079999994, 2),
 (datetime.date(2021, 3, 1), 82569.71100000001, 272037.3854999999, 2),
 (datetime.date(2021, 3, 22), 138519.70599999995, 272031.87113333336, 2),
 (datetime.date(2021, 3, 29), 150299.74899999998, 271685.5309, 2),
 (datetime.date(2021, 1, 25), 305789.77599999984, 271679.73487499997, 2),
 (datetime.date(2021, 3, 23), 151389.65200000003, 271433.20119999995, 2),
 (datetime.date(2021, 2, 22), 254979.73199999993, 270898.7259333333, 2),
 (datetime.date(2021, 2, 28), 266309.75099999993, 270796.38603333326, 2),
 (datetime.date(2021, 3, 26), 180879.57499999987, 270596.0356333333, 2),
 (datetime.date(2021, 2, 15), 198879.6419999999, 270448.0611666666, 2),
 (datetime.date(2021, 2, 13), 213979.68799999994, 270214.23056666664, 2),
 (datetime.date(2021, 3, 12), 471189.6739999998, 269760.37883333326, 2),
 (datetime.date(2021, 2, 23), 256189.672, 269205.72549999994, 2),
 (datetime.date(2021, 3, 25), 312604.7569999999, 268715.5328, 2),
 (datetime.date(2021, 1, 24), 351809.736, 268195.8217826087, 2),
 (datetime.date(2021, 2, 16), 188944.79199999996, 268135.8888666666, 2),
 (datetime.da

## Resumen

En este capítulo, aprendiste sobre las funciones de ventana, que generan resultados para una fila en función de su posición dentro del conjunto de datos o subgrupos dentro del conjunto de datos. Esto difiere de las funciones simples que aprendiste en el Capítulo 3, SQL para la Preparación de Datos, que generan un resultado para una fila independientemente de las características del conjunto de datos, y difiere de las funciones de agregación que aprendiste en el Capítulo 4, Funciones de Agregación para el Análisis de Datos, que generan un resultado para todas las filas en un conjunto de datos o subgrupos en el conjunto de datos.
Aprendiste algunas de las funciones de ventana más comunes, incluyendo COUNT, SUM y RANK. También aprendiste cómo construir una ventana básica utilizando la cláusula OVER. La salida de la función de ventana depende de la posición actual de la fila en el conjunto de datos o subgrupos dentro del conjunto de datos, lo que se llama partición, así como de la colección de filas requeridas por el cálculo, que se llama ventana. Por lo tanto, hay varias palabras clave que pueden afectar cómo se realiza el cálculo, como PARTITION BY, ORDER BY y las palabras clave de marco de ventana. La cláusula PARTITION BY determina la partición, la cláusula ORDER BY determina la posición de la fila dentro de la partición y las palabras clave del marco de ventana determinan el rango y el tamaño de la ventana. Luego aprendiste cómo utilizar las funciones de ventana para obtener información analítica. Por ejemplo, al definir un marco de ventana sobre un resumen diario, como las ventas diarias, puedes crear estadísticas móviles y obtener información útil sobre la tendencia temporal de las ventas.
En este punto, has aprendido todas las declaraciones fundamentales de SQL. Has aprendido cómo manejar el ciclo completo de CRUD utilizando SQL, cómo combinar tablas utilizando JOIN y UNION, y has aprendido cómo utilizar diferentes tipos de funciones para obtener los resultados deseados. En el Capítulo 6, Importación y Exportación de Datos, verás cómo importar y exportar datos para utilizar SQL con otros programas. Utilizarás el comando COPY para cargar datos en tu base de datos de manera masiva. También utilizarás Excel para procesar datos de tu base de datos y luego simplificarás tu código utilizando Python.